<div style="font-size: 30px;font-family: Times New Roman;color: #e12eb0"><em><strong>Introduction Générale</strong></em></div>

<p>Le nettoyage de données est l'opération de détection et de correction (ou suppression) d'erreurs présentes sur des données stockées dans des bases de données ou dans des fichiers.

Le nettoyage de données est un des problèmes majeurs des entrepôts de données.

Les données présentes dans les bases de données peuvent avoir plusieurs types d'erreurs comme des erreurs de frappe, des informations manquantes, des imprécisions etc. La partie impropre de la donnée traitée peut être remplacée, modifiée ou supprimée. Le processus de nettoyage identifie les données erronées et les corrige automatiquement avec un programme informatique ou les propose à un humain pour qu'il effectue les modifications.

Le nettoyage de données est différent de la validation de données. La validation de données est l'étape qui consiste à vérifier et rejeter les données qui ne respectent pas certaines règles avant l'ajout en base de données, alors que le nettoyage intervient après (sur des données déjà présentes en base de données).
 </p>

<div style="font-size: 30px;font-family: Times New Roman;color: #2f8ede"><em><strong>Partie I</strong></em></div>

<p><strong>Importer les bibliothèques numpy et pandas et importer functools de reduce</strong><p>

In [1]:
# Importer les bibliothèques nécessaires
import numpy as np
import pandas as pd
from functools import reduce 

<div style="font-size: 20px;font-family: Times New Roman;color: #e79f82"><strong>Remarques</strong></div>

<p><strong>Pandas</strong> is a powerhouse tool that allows you to do anything and everything with tabular or columnar data sets -- analyzing, organizing, sorting, filtering, aggregating, cleaning, calculating, and more!</p>

<p><strong>Matplotlib</strong> is an amazing visualization library in Python for 2D plots of arrays. Matplotlib is a multi-platform data visualization library built on NumPy arrays. One of the greatest benefits of visualization is that it allows us visual access to huge amounts of data in easily digestible visuals. Matplotlib consists of several plots like line, bar, scatter, histogram etc.</p>

<p><strong>functools</strong>Le module functools est utilisé pour des fonctions de haut niveau : des fonctions qui agissent sur ou revoient d'autres fonctions. En général, tout objet appelable peut être utilisé comme une fonction pour les besoins de ce module.</p>

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>1 Exploration du dataset: "BL-Flickr-Images-Book" dataset</strong></div>

In [2]:
df = pd.read_csv('BL-Flickr-Images-Book.csv')
#affocher les 5 premiers lignes
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

--> La première étape consiste à importer les données. Souvent, Ces données dans des fichiers CSV (valeurs séparées par des virgules) .

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>2 Jeter un coup d'oeil sur DataSet</strong></div>

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8287 entries, 0 to 8286
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Identifier              8287 non-null   int64  
 1   Edition Statement       773 non-null    object 
 2   Place of Publication    8287 non-null   object 
 3   Date of Publication     8106 non-null   object 
 4   Publisher               4092 non-null   object 
 5   Title                   8287 non-null   object 
 6   Author                  6509 non-null   object 
 7   Contributors            8287 non-null   object 
 8   Corporate Author        0 non-null      float64
 9   Corporate Contributors  0 non-null      float64
 10  Former owner            1 non-null      object 
 11  Engraver                0 non-null      float64
 12  Issuance type           8287 non-null   object 
 13  Flickr URL              8287 non-null   object 
 14  Shelfmarks              8287 non-null   

<p>Dataset consiste de 8287 quartiers et de 15 attributs : Identifier, Edition Statement, Place of Publication, Date of Publication, Date of Publication, Publisher,Title, Author, Contributors, Corporate Author, Corporate Contributors, Former owner, Engraver , Issuance type, Flickr URL ,Shelfmarks.  
Il y'a aussi des données manquantes telle que on vérifier par:</p>

In [4]:
missing = df.isna().sum()
print(missing)

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64


<div style="font-size: 20px;font-family: Times New Roman;color: #e79f82"><strong>Attention</strong></div>

<p><strong>.isna().sum():</strong>Compter le total des valeurs manquantes dans un dataframe</p>


<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>3 Dropping unnecessary columns</strong></div>

<p>Nous allons restreindre l'étude à seulement 7 colonnes pour cela nous allons supprimer 7 colonnes que nous jugeons unitiles por le moment :
</p>

In [5]:
#Eliminer les colonnes dans la liste de to_drop du dataframe df
to_drop= ['Edition Statement','Corporate Author','Corporate Contributors','Former owner','Engraver','Contributors','Issuance type','Shelfmarks']
df.drop(to_drop,1,inplace=True)
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                         Flickr URL  
0      A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
1  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
2  A., A. A.  http://www.flickr.com/photos/britishlibrary/ta...  
3  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...  
4  A., E. S.  http://www.flickr.com/photos/britishlibrary/ta...

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>4 Setting the index of the dataset</strong></div>

In [6]:
### Changerl'index
df.set_index('Identifier', inplace=True)

Dans cette étape on prend l'identifiant comme index car on a besoin seulement soit de l'identifiant soit du l'index.

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>5 Donner un appérçu sur la colonne Date of Publication</strong></div>

In [7]:
df['Date of Publication'].head(25)

Identifier
206            1879 [1878]
216                   1868
218                   1869
472                   1851
480                   1857
481                   1875
519                   1872
667                    NaN
874                   1676
1143                  1679
1280                  1802
1808                  1859
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
Name: Date of Publication, dtype: object

<p><center><strong>Remarque : Observer la date de publication qui correspond à l'index 206. Ca nécessite un nettoyage et une correction : il faut éliminer certains symboles</strong></center></p>

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>6 Cleaning columns using the .apply function</strong></div>

<div style="font-size: 18px;font-family: Times New Roman;color: #4816b8">Ecrire une fonction qui permet de nettoyer la colonne des dates</div>

In [53]:
unwanted_characters= ['[', ',', '-']

def clean_dates(item):
    dop=str(item.loc['Date of Publication'])
    if dop=='nan' or dop[0] =='[':
        return 'NaN' 
    for character in unwanted_characters:
        if character in dop:
            character_index=dop.find(character)
            dop=dop[:character_index]
    return dop
df['Date of Publication'] =df.apply(clean_dates, axis=1)

In [54]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

<div style="font-size: 18px;font-family: Times New Roman;color: #4816b8">Deuxième Méthode:</div>

In [55]:
#exécuter pour voir le résultat
unwanted_characters = ['[', ',', '-']

def clean_dates(dop):
    dop = str(dop)
    if dop.startswith('[') or dop == 'nan':
        return 'NaN'
    for character in unwanted_characters:
        if character in dop:
            character_index = dop.find(character)
            dop = dop[:character_index]
    return dop

df['Date of Publication'] = df['Date of Publication'].apply(clean_dates)

In [56]:
df.head(10)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London               1879    
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   

                                                        Title         Author  \
Identifier                                                                     
206                         Walter Forbes. [A novel.] By A. A          A. A.   
216         All for Greed. [A novel. The dedication signed...      A., A. A.   
218         Love the Avenger. By the author of “All for Gr...      A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...      A., E. S.   
480         [The World in which I live, and my place in it...      A., E. S.   
481         [The World in which I live, and my place in it...      A., E. S.   
519         Lagonells. By the author of Darmayne (F. E. A....      A., F. E.   
667         The Coming of Spring, and other poems. By J. A...  A., J.|A., J.   
874         A Warning to the inhabitants of England, and L...         Remaʿ.   
1143        A Satyr against Vertue. (A poem: supposed to b...         A., T.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
481         http://www.flickr.com/photos/britishlibrary/ta...  
519         http://www.flickr.com/photos/britishlibrary/ta...  
667         http://www.flickr.com/photos/britishlibrary/ta...  
874         http://www.flickr.com/photos/britishlibrary/ta...  
1143        http://www.flickr.com/photos/britishlibrary/ta...

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>7 Observer la collone Author</strong></div>

<div style="font-size: 18px;font-family: Times New Roman;color: #4816b8">Completer la fonction clean_author_namesqui permet de corriger les valeurs de cet colonne</div>

In [57]:
# compléter la fonction suivante ( à la place de pass)
def clean_author_names(author):
    author = str(author)
    if author == 'nan':
        return 'NaN'
    author = author.split(',')
    if len(author) == 1:
        name = filter(lambda x: x.isalpha(), author[0])
        return reduce(lambda x, y: x + y, name)
    last_name, first_name = author[0], author[1]
    first_name = first_name[:first_name.find('-')] if '-' in first_name else first_name

    if first_name.endswith(('.', '.|')):
        parts = first_name.split('.')
        
        if len(parts) > 1:
            first_occurence = first_name.find('.')
            final_occurence = first_name.find('.', first_occurence + 1)
            first_name = first_name[:final_occurence]
        else:
            first_name = first_name[:first_name.find('.')]
    
    last_name = last_name.capitalize()
    return f'{first_name} {last_name}'

df['Author'] = df['Author'].apply(clean_author_names)
df.head(10)

Place of Publication Date of Publication  \
Identifier                                                            
206                                      London               1879    
216                    London; Virtue & Yorston                1868   
218                                      London                1869   
472                                      London                1851   
480                                      London                1857   
481                                      London                1875   
519                                      London                1872   
667         pp. 40. G. Bryan & Co: Oxford, 1898                 NaN   
874                                     London]                1676   
1143                                     London                1679   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   
481             William Macintosh   
519                    The Author   
667                           NaN   
874                           NaN   
1143                          NaN   

                                                        Title    Author  \
Identifier                                                                
206                         Walter Forbes. [A novel.] By A. A        AA   
216         All for Greed. [A novel. The dedication signed...   A. A A.   
218         Love the Avenger. By the author of “All for Gr...   A. A A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...   E. S A.   
480         [The World in which I live, and my place in it...   E. S A.   
481         [The World in which I live, and my place in it...   E. S A.   
519         Lagonells. By the author of Darmayne (F. E. A....   F. E A.   
667         The Coming of Spring, and other poems. By J. A...   J.|A A.   
874         A Warning to the inhabitants of England, and L...     Remaʿ   
1143        A Satyr against Vertue. (A poem: supposed to b...      T A.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...  
481         http://www.flickr.com/photos/britishlibrary/ta...  
519         http://www.flickr.com/photos/britishlibrary/ta...  
667         http://www.flickr.com/photos/britishlibrary/ta...  
874         http://www.flickr.com/photos/britishlibrary/ta...  
1143        http://www.flickr.com/photos/britishlibrary/ta...

<div style="font-size: 20px;font-family: Times New Roman;color: #e6005c"><strong>8 Observer la colonne title</strong></div>

<div style="font-size: 18px;font-family: Times New Roman;color: #4816b8">Completer la fonction clean_title qui permet de corriger les valeurs de cet colonne</div>

In [58]:
def clean_title(title):
    
    if title == 'nan':
        return 'NaN'
    
    if title[0] == '[':
        title = title[1: title.find(']')]
        
    if 'by' in title:
        title = title[:title.find('by')]
    elif 'By' in title:
        title = title[:title.find('By')]
        
    if '[' in title:
        title = title[:title.find('[')]

    title = title[:-2]
        
    title = list(map(str.capitalize, title.split()))
    return ' '.join(title)
    
df['Title'] = df['Title'].apply(clean_title)
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London               1879    
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title    Author  \
Identifier                                                                
206                                             Walter Forbes        AA   
216                                             All For Greed   A. A A.   
218                                          Love The Avenger   A. A A.   
472         Welsh Sketches, Chiefly Ecclesiastical, To The...   E. S A.   
480             The World In Which I Live, And My Place In It   E. S A.   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...